pkg/autoupdate/findnew.go

```diff
@@ -266,6 +263,41 @@ func getUpdateDir(fullBinaryPath string) string {
 	return fmt.Sprintf("%s%s", components[0], updateDirSuffix)
 }
 
+// Find the possible updates. filepath.Glob returns a list of things
+// that match the requested pattern. We sort the list to ensure that
+// we can tell which ones are earlier or later (remember, these are
+// timestamps).
+func getPossibleUpdates(updateDir, binaryName string) ([]string, error) {
+	// If this is launcher running on macOS, then we should have app bundles available instead --
+	// check for those first.
+	if runtime.GOOS == "darwin" {
+		binarySuffix := filepath.Join("Contents", "MacOS", binaryName)
+		fileGlob := filepath.Join(updateDir, "*", "*.app", binarySuffix)
+		possibleUpdates, err := filepath.Glob(fileGlob)
+
+		if err == nil && len(possibleUpdates) > 0 {
+			appBundleNames := make([]string, len(possibleUpdates))
+			for i, binaryPath := range possibleUpdates {
+				appBundleNames[i] = strings.TrimSuffix(binaryPath, "/"+binarySuffix)
+			}
+			sort.Strings(appBundleNames)
+			return appBundleNames, nil
+		}
+	}
+
+	// Either not macOS/launcher or no app bundles found. Fall back to searching for binaries.
```

Comment:
Total speculation, but I wonder if it's any cleaner to just return all the paths (app and not) in the desired sort order.

Reply:
So, I'd spent some time doing it this way this morning, and the implementation wasn't significantly cleaner one way or the other. The reason I ended up keeping it like this (returning early if we've found suitable app bundle binaries) is because if we put them all (app and not) in together and sort them, then in the tests, we end up always choosing the top-level binary paired with the app bundle instead of the binary inside the app bundle (`<update-dir>/<timestamp>/binary` is always alphabetically before `<update-dir>/<timestamp>/Test.app/Contents/MacOS/binary`). Fixing that nicely would've required another edge-case-for-app-bundles carveout, and it felt a little brittle to just rely on the name of the app bundle never changing to be after `launcher` in the alphabet.

This also covers why I didn't end up doing mixed app bundle and non-app bundle for the tests like you mention [here](https://github.com/kolide/launcher/pull/932#pullrequestreview-1179716935).

However, this does leave us in the position where if `launcher-updates` has a couple binary-only updates that are fine, but some app bundle updates that are broken, we won't fall back to the binary-only updates because we're only looking for the app bundle updates. I'm not sure how big of a deal this is. I could fix it fairly simply by throwing the `checkExecutable` call in here, but it means more calls to `checkExecutable` than we technically need.

Let me know what you think or if it's easier to chat about it briefly instead.

Reply:
I think, like above, this sounds like a tradeoff between a larger refactoring and small PR.

When you explain it, I immediately think we could move `checkExecutable` in here, and skip it in the caller. And then I realize we could put the reverse sort here, and return the first thing with a valid check. But I guess that bumps into the cleanup routine....

_Anyhow_  I think this is fine, thank you for explaining, and the downside you mention feels like an acceptable tradeoff to me.